src/drivers/imx/interrupt.c

```diff
@@ -3,26 +3,413 @@
+}
+
+static inline void irqstr_update_bits(uint32_t reg, uint32_t mask,
+				      uint32_t value)
+{
+	io_reg_update_bits(IRQSTR_BASE_ADDR + reg, mask, value);
+}
+
+/* IRQ_STEER helper methods
+ * These methods are usable in any IRQ_STEER driver, not specific to SOF
+ */
+
+static void irqstr_enable_hw(void)
+{
+	irqstr_write(IRQSTR_CHANCTL, 1);
+}
+
+static void irqstr_disable_hw(void)
+{
+	irqstr_write(IRQSTR_CHANCTL, 0);
+}
+
+/* irqstr_get_status_word() - Get an interrupt status word
+ * @index The index of the status word
+ *
+ * Get the status of interrupts 32*index .. 32*(index+1)-1 in a word.
+ * This status is in one hardware register.
+ * Return: Status register word for the corresponding interrupts
+ */
+static uint32_t irqstr_get_status_word(uint32_t index)
+{
+	/* First 32 interrupts are reserved, we just give 0 */
+	if (!index)
+		return 0;
+	/* On out of range for our platform, be silent */
+	if (index > 15)
+		return 0;
+	return irqstr_read(IRQSTR_CH_STATUS(index));
+}
+
+/* Mask, that is, disable interrupts */
+static void irqstr_mask_int(uint32_t irq)
+{
+	uint32_t mask;
+
+	if (irq < IRQSTR_RESERVED_IRQS_NUM || irq >= IRQSTR_IRQS_NUM)
+		return; // Unusable interrupts
+	mask = IRQSTR_INT_MASK(irq);
+	irqstr_update_bits(IRQSTR_CH_MASK(IRQSTR_INT_REG(irq)), mask, 0);
+}
+
+/* Unmask, that is, enable interrupts */
+static void irqstr_unmask_int(uint32_t irq)
+{
+	uint32_t mask;
+
+	if (irq < IRQSTR_RESERVED_IRQS_NUM || irq >= IRQSTR_IRQS_NUM)
+		return; // Unusable interrupts
+	mask = IRQSTR_INT_MASK(irq);
+	irqstr_update_bits(IRQSTR_CH_MASK(IRQSTR_INT_REG(irq)), mask, mask);
+}
+
+/* SOF specific part of the driver */
+
+/* Quirk of the driver in SOF:
+ * -> Interrupts 0-31 are hardware
+ * -> Interrupts 32-63 are unusable, as they are reserved in irqstr. We
+ *  will never get an event on these shared interrupt lines.
+ * -> Interrupts 64-543 are usable, mapping to 32-512 in IRQSTR itself
+ * The above functions expect the 32-512 interrupts valid, not the
+ * shifted SOF ones.
+ */
+
+const char * const irq_name_irqsteer[] = {
+	"irqsteer0",
+	"irqsteer1",
+	"irqsteer2",
+	"irqsteer3",
+	"irqsteer4",
+	"irqsteer5",
+	"irqsteer6",
+	"irqsteer7"
+};
+
+#define IRQ_MAX_TRIES	1000
+
+/* Extract the 64 status bits corresponding to output interrupt line
+ * index (64 input interrupts)
+ */
+static uint64_t get_irqsteer_interrupts(uint32_t index)
+{
+	uint64_t result = irqstr_get_status_word(2 * index + 1);
+
+	result <<= 32;
+	result |= irqstr_get_status_word(2 * index);
+	return result;
+}
+
+/**
+ * get_first_irq() Get the first IRQ bit set in this group.
+ * @ints The 64 input interrupts
+ *
+ * Get the first pending IRQ in the group. For example, get_first_irq(0x40)
+ * will return 6 (as 1 << 6 is 0x40), while get_first_irq(0) will return -1.
+ *
+ * Return: -1 if all interrupts are clear, or a shift value if at least
+ * one interrupt is set.
+ */
+static int get_first_irq(uint64_t ints)
+{
+	return __builtin_ffsll(ints) - 1;
```

Comment:
Could you define it in sof/common.h, so we don't need to use builtin directly?
`#define ffsll(i) __builtin_ffsll(i)`